### PR TITLE
Bugfix for a Kratos Parameters Method (recursively check defaults)

### PR DESCRIPTION
**Pull request by maceligueta**
A test has been added too. Also a necessary method to check the correctness.

### CODE REVIEW

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h:697]: shouldn't you check only for the type here? do you need it to be identical?

File: kratos/tests/test_kratos_parameters.py
Comment (RiccardoRossi) [inline kratos/tests/test_kratos_parameters.py]: i think you can completely remove the print

**Review by RiccardoRossi — COMMENTED** [review]
+1 - i addes some comments, but they are mostly for discussion

**Review by RiccardoRossi — APPROVED** [review]

**Comment by philbucher** [comment]
Nice improvement!

File: kratos/includes/kratos_parameters.h
Comment (maceligueta) [reply kratos/includes/kratos_parameters.h:697]: I think the method does what it says (Equivalent), and I would leave it like this. But I agree that it would be useful to add another method (very similar) that only checks the type, after modifications have been done by other processes.

**Comment by maceligueta** [comment]
Please do not merge yet, I will add a new method according to @RiccardoRossi 's comment.

**Comment by maceligueta** [comment]
@RiccardoRossi : can you check if the new method does what you expected? Just checks the type, but the old method is kept, as it looks useful for testing.

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: I would give an output message here telling what is different. what do you think?

File: kratos/tests/test_kratos_parameters.py
Comment (RiccardoRossi) [inline kratos/tests/test_kratos_parameters.py:299]: You are never checking for the False case in HasSameJeyAndTypeOfValuesAs

**Review by RiccardoRossi — COMMENTED** [review]
your code looks fine to me, apart for the comments i am putting in the code, which resume to

1 - you should output which var has a type mismatch
2 - you are not checking for a false case in the tests

other than that +1

File: kratos/tests/test_kratos_parameters.py
Comment (maceligueta) [reply kratos/tests/test_kratos_parameters.py:299]: I added an extra assertFalse in the las commit

File: kratos/includes/kratos_parameters.h
Comment (maceligueta) [reply kratos/includes/kratos_parameters.h]: I put it, but then I commented it out. It would print the message everytime we ran the tests...
Considering that this function is mainly used for testing, I would leave it without the message. However, you can uncomment it easily if you think it's better uncommented.

**Comment by RiccardoRossi** [comment]
@maceligueta i added a function IsMatrix and GetMatrix (still missing a SetMatrix)

those funcions should help with  #729 so maybe @peterjwilson or @philbucher  might have time to complete the funcitonality...

missing things are:
SetMatrix
IsVector ---- these ones giving back directly a kratos vector
GetVector
SetVector

also the test i made does not check yet for the errors being thrown

**Comment by pooyan-dadvand** [comment]
👍

**Comment by philbucher** [comment]
@RiccardoRossi I added the requested functions including the tests.
Could you please check them?
Especially the `SetVector` and `SetMatrix` methods. Search for "Riccardo" to find my questions

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: We should not throw an error, simply return false

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: Return false

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: We shoildnreturn here. Not throw an error

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: Well we could return a vector even of size 1. It is relevant in some corner cases

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h]: To be honest I don't know. Look into the rapidjson documentation, or ask in the relevant list (I would have to do the same)

**Comment by philbucher** [comment]
Hi @RiccardoRossi I implemented your requested changes:
I have four small questions:
1. In the `SetVector` and `SetMatrix` Methods, how to pass the `Vector`/`Matrix`? Right now I do it by const reference, but I think this is not correct.
2. Should the  `IsVector` and `IsMatrix` Methods return `false` for an empty list (`[]` or`[[]]`) Right now they would return `true` in this case.
3. I refractored the code to use the new `KRATOS_ERROR` macros, is this ok for you? I did this in a separate commit, so we can easily revert it.
4. While doing this, I noticed that `GetBool` behaves differently from other `Get` functions. Basically if the value is not a bool, it calls `RecursivelyFindValue`, but still throws an error afterwards. Is this desired?

**Comment by philbucher** [comment]
I answered Point 4 myself, this functions gave a segfault, so i unified it
Also I unified the test for `IsXXX` and `GetXXX`
@RiccardoRossi please let me know if you want changes

File: kratos/includes/kratos_parameters.h
Comment (philbucher) [inline kratos/includes/kratos_parameters.h:280]: This gives a segfault if the value is not a boolean

**Comment by maceligueta** [comment]
@RiccardoRossi , you approved this PR long ago, but we would need you to say OK after all the changes that were made.

File: kratos/includes/kratos_parameters.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h:411]: Const reference looks ok to me. Why do you think it is wrong? Anyway you are copying it to the is In, at least by passing a reference you avoid copying it twice

**Review by RiccardoRossi — APPROVED** [review]
It all looks fine to me. Sorry for the delay in reviewing

**Comment by maceligueta** [comment]
@philbucher : I would merge now, but I let you do it, because you made the last changes and Riccardo asked you something. As soon as you feel it, merge! Thanks

**Pull request merged**
